lib/puppet/ssl/ssl_provider.rb

```diff
@@ -71,6 +74,30 @@ def create_system_context(cacerts:, path: Puppet[:ssl_trust_store])
       end
     end
 
+    if include_client_cert
+      cert_provider = Puppet::X509::CertProvider.new
+      private_key = cert_provider.load_private_key(Puppet[:certname], required: false)
+      client_cert = cert_provider.load_client_cert(Puppet[:certname], required: false)
+
+      if private_key && client_cert
+        client_chain = verify_cert_with_store(store, client_cert)
+
+        if !private_key.is_a?(OpenSSL::PKey::RSA) && !private_key.is_a?(OpenSSL::PKey::EC)
+          raise Puppet::SSL::SSLError, _("Unsupported key '%{type}'") % { type: private_key.class.name }
+        end
+
+        unless client_cert.check_private_key(private_key)
+          raise Puppet::SSL::SSLError, _("The certificate for '%{name}' does not match its private key") % { name: subject(client_cert) }
+        end
```

Comment:
refactor with existing code

Reply:
Updated in https://github.com/puppetlabs/puppet/pull/8908/commits/e4bb4733ceb44558e904c7b5191594217c19dd1e